File: flutter_application_2/lib/services/ai_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/ai_model.dart';
import '../models/lesson_plan.dart';

/// 内容类型枚举，表示可以生成的内容类型
enum ContentType {
  text,
  image,
  audio,
  lessonPlan,
  quiz,
  presentation
}

/// 内容增强类型枚举
enum EnhancementType {
  simplify,
  elaborate,
  visualize,
  translate,
  adapt
}

/// AI服务类，负责与多模态大模型API交互
class AIService {
  final String _baseUrl;
  final http.Client _client;
  
  AIService({
    required String baseUrl,
    http.Client? client,
  }) : _baseUrl = baseUrl,
       _client = client ?? http.Client();
  
  /// 生成教案
  /// 
  /// [topic] 教案主题
  /// [grade] 年级
  /// [subject] 学科
  /// [durationMinutes] 课程时长（分钟）
  /// [model] 使用的AI模型
  Future<LessonPlan> generateLessonPlan({
    required String topic,
    required String grade,
    required String subject,
    required int durationMinutes,
    required AIModel model,
  }) async {
    try {
      // 这里是模拟API调用，实际项目中应替换为真实API调用
      // await Future.delayed(const Duration(seconds: 2));
      
      final response = await _client.post(
        Uri.parse('$_baseUrl/generate/lesson-plan'),
        headers: {
          'Content-Type': 'application/json',
        },
        body: jsonEncode({
          'topic': topic,
          'grade': grade,
          'subject': subject,
          'durationMinutes': durationMinutes,
          'model': model.toString(),
        }),
      );
      
      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return LessonPlan.fromJson(data);
      } else {
        throw Exception('生成教案失败: ${response.statusCode}');
      }
    } catch (e) {
      // 模拟数据，实际项目中应删除
      return _getMockLessonPlan(topic, grade, subject, durationMinutes);
    }
  }
  
  /// 生成多模态内容
  /// 
  /// [prompt] 提示词
  /// [type] 内容类型
  /// [model] 使用的AI模型
  Future<Map<String, dynamic>> generateContent({
    required String prompt,
    required ContentType type,
    required AIModel model,
  }) async {
    // 实际项目中应实现真实API调用
    await Future.delayed(const Duration(seconds: 2));
    
    // 返回模拟数据
    return {
      'id': DateTime.now().millisecondsSinceEpoch.toString(),
      'type': type.toString(),
      'content': '这是基于提示"$prompt"生成的${_getContentTypeName(type)}内容',
      'createdAt': DateTime.now().toIso8601String(),
    };
  }
  
  /// 增强内容
  /// 
  /// [content] 原始内容
  /// [enhancementType] 增强类型
  Future<Map<String, dynamic>> enhanceContent({
    required Map<String, dynamic> content,
    required EnhancementType enhancementType,
  }) async {
    // 实际项目中应实现真实API调用
    await Future.delayed(const Duration(seconds: 1));
    
    // 返回模拟数据
    return {
      ...content,
      'content': '${content['content']} (已${_getEnhancementTypeName(enhancementType)})',
      'updatedAt': DateTime.now().toIso8601String(),
    };
  }
  
  /// 图像转文本
  Future<String> imageToText(String imagePath) async {
    // 实际项目中应实现真实API调用
    await Future.delayed(const Duration(seconds: 1));
    return '这是从图片中提取的文本内容';
  }
  
  /// 文本转语音
  Future<String> textToSpeech(String text) async {
    // 实际项目中应实现真实API调用
    await Future.delayed(const Duration(seconds: 1));
    return 'audio_file_url.mp3';
  }
  
  // 获取内容类型名称
  String _getContentTypeName(ContentType type) {
    switch (type) {
      case ContentType.text:
        return '文本';
      case ContentType.image:
        return '图片';
      case ContentType.audio:
        return '音频';
      case ContentType.lessonPlan:
        return '教案';
      case ContentType.quiz:
        return '测验';
      case ContentType.presentation:
        return '演示文稿';
    }
  }
  
  // 获取增强类型名称
  String _getEnhancementTypeName(EnhancementType type) {
    switch (type) {
      case EnhancementType.simplify:
        return '简化';
      case EnhancementType.elaborate:
        return '详细阐述';
      case EnhancementType.visualize:
        return '可视化';
      case EnhancementType.translate:
        return '翻译';
      case EnhancementType.adapt:
        return '适应性调整';
    }
  }
  
  // 生成模拟教案数据
  LessonPlan _getMockLessonPlan(String topic, String grade, String subject, int durationMinutes) {
    return LessonPlan.create(
      title: '$topic - $grade $subject 教案',
      subject: subject,
      grade: grade,
      durationMinutes: durationMinutes,
      objectives: '1. 理解$topic的基本概念\n2. 掌握$topic的应用方法\n3. 能够解决相关问题',
      materials: '教科书、多媒体设备、学习单',
      assessment: '课堂提问、小组活动表现、课后作业',
      notes: '注意引导学生主动思考，鼓励参与讨论',
      sections: [
        LessonSection(
          title: '导入',
          content: '通过日常生活例子引入$topic话题，激发学生兴趣',
          durationMinutes: 5,
          activityType: '讲解',
        ),
        LessonSection(
          title: '新知识讲解',
          content: '详细讲解$topic的核心概念和原理',
          durationMinutes: 15,
          activityType: '讲解',
        ),
        LessonSection(
          title: '小组活动',
          content: '学生分组讨论$topic的应用场景',
          durationMinutes: 10,
          activityType: '小组活动',
        ),
        LessonSection(
          title: '练习与巩固',
          content: '完成相关习题，巩固所学知识',
          durationMinutes: 10,
          activityType: '个人练习',
        ),
        LessonSection(
          title: '总结与反思',
          content: '总结课堂内容，布置课后作业',
          durationMinutes: 5,
          activityType: '讲解',
        ),
      ],
    );
  }
} 

File: flutter_application_2/backend/app/services/ai_service.py
import json
import httpx
import time
import uuid
from typing import Dict, Any, Optional, List
from fastapi import HTTPException, UploadFile

from app.core.config import (
    AI_SERVICE_URL, AI_SERVICE_KEY,
    TEXT_CHAT_API_URL, TEXT_CHAT_API_KEY,
    LESSON_PLAN_API_URL, LESSON_PLAN_API_KEY,
    CONTENT_CONVERT_API_URL, CONTENT_CONVERT_API_KEY,
    IMAGE_RECOGNITION_API_URL, IMAGE_RECOGNITION_API_KEY,
    IMAGE_GENERATION_API_URL, IMAGE_GENERATION_API_KEY
)
from app.models.ai import (
    LessonPlanGenRequest, 
    ContentConvertRequest, 
    ChatRequest,
    ImageGenerationRequest
)
from app.models.lesson_plan import LessonPlan, LessonSection

# 模拟数据，实际项目应使用数据库
ai_tasks = {}

class AIService:
    @staticmethod
    async def chat_with_text(request: ChatRequest) -> Dict[str, Any]:
        """
        调用AI服务进行文本对话
        """
        try:
            # 创建任务ID
            task_id = str(uuid.uuid4())
            
            # 实际项目中，调用真实的AI API
            # 以下注释部分为实际调用示例代码
            """
            async with httpx.AsyncClient(timeout=60.0) as client:
                response = await client.post(
                    TEXT_CHAT_API_URL,
                    headers={"Authorization": f"Bearer {TEXT_CHAT_API_KEY}"},
                    json={
                        "messages": [{"role": msg.role, "content": msg.content} for msg in request.messages],
                        "model": request.model,
                        "temperature": request.temperature,
                        "max_tokens": request.max_tokens
                    }
                )
                response.raise_for_status()
                result = response.json()
            """
            
            # 模拟回复数据
            last_message = request.messages[-1].content if request.messages else ""
            result = {
                "id": f"chatcmpl-{task_id}",
                "object": "chat.completion",
                "created": int(time.time()),
                "model": request.model or "gpt-3.5-turbo",
                "choices": [
                    {
                        "message": {
                            "role": "assistant",
                            "content": f"这是对您消息的模拟回复。您的问题是：{last_message}"
                        },
                        "finish_reason": "stop",
                        "index": 0
                    }
                ]
            }
            
            # 存储任务
            ai_tasks[task_id] = {
                "type": "chat",
                "status": "completed",
                "request": request.dict(),
                "result": result
            }
            
            return result
            
        except Exception as e:
            print(f"文本对话出错: {str(e)}")
            raise HTTPException(status_code=500, detail=f"文本对话失败: {str(e)}")

    @staticmethod
    async def generate_lesson_plan(request: LessonPlanGenRequest) -> LessonPlan:
        """
        调用AI服务生成教案
        """
        try:
            # 创建任务ID
            task_id = str(uuid.uuid4())
            
            # 实际项目中，应使用httpx调用真实的AI API
            # 以下注释部分为实际调用示例代码
            """
            async with httpx.AsyncClient(timeout=60.0) as client:
                response = await client.post(
                    LESSON_PLAN_API_URL,
                    headers={"Authorization": f"Bearer {LESSON_PLAN_API_KEY}"},
                    json=request.dict()
                )
                response.raise_for_status()
                result = response.json()
                
                # 将API返回的结果转换为LessonPlan对象
                lesson_plan = LessonPlan(**result)
                return lesson_plan
            """
            
            # 模拟教案生成
            lesson_plan = AIService._mock_lesson_plan(
                topic=request.topic,
                grade=request.grade,
                subject=request.subject,
                duration_minutes=request.duration_minutes
            )
            
            # 存储结果
            ai_tasks[task_id] = {
                "type": "lesson_plan",
                "status": "completed",
                "request": request.dict(),
                "result": lesson_plan.dict()
            }
            
            return lesson_plan
            
        except Exception as e:
            # 记录错误
            print(f"生成教案出错: {str(e)}")
            raise HTTPException(status_code=500, detail=f"生成教案失败: {str(e)}")
    
    @staticmethod
    async def convert_content(request: ContentConvertRequest, file: Optional[UploadFile] = None) -> Dict[str, Any]:
        """
        调用AI服务进行多模态内容转换
        """
        try:
            # 创建任务ID
            task_id = str(uuid.uuid4())
            
            # 实际项目中，应调用真实的AI API
            # 以下注释部分为实际调用示例代码
            """
            # 准备请求数据
            data = {"type": request.type}
            files = {}
            
            if request.content:
                data["content"] = request.content
                
            if file:
                # 如果有上传文件，添加到请求中
                file_content = await file.read()
                files = {"file": (file.filename, file_content, file.content_type)}
            elif request.file_url:
                # 如果提供了文件URL，添加到请求中
                data["file_url"] = request.file_url
                
            async with httpx.AsyncClient(timeout=60.0) as client:
                response = await client.post(
                    CONTENT_CONVERT_API_URL,
                    headers={"Authorization": f"Bearer {CONTENT_CONVERT_API_KEY}"},
                    data=data,
                    files=files
                )
                response.raise_for_status()
                result = response.json()
            """
            
            # 根据转换类型生成不同的模拟结果
            result = {}
            if request.type == "text2img":
                # 模拟生成图片URL
                result = {
                    "url": f"/mock/images/{task_id}.png",
                    "width": 512,
                    "height": 512
                }
            elif request.type == "img2text":
                # 模拟图片识别结果
                result = {
                    "text": "这是一张图片的识别结果。实际使用时将返回图像的真实描述。"
                }
            elif request.type == "speech2text":
                # 模拟语音识别结果
                result = {
                    "text": "这是一段语音的转写结果。实际使用时将返回语音的真实转写内容。"
                }
            elif request.type == "text2speech":
                # 模拟语音合成结果
                result = {
                    "url": f"/mock/audio/{task_id}.mp3",
                    "duration": 12.5
                }
            else:
                raise ValueError(f"不支持的转换类型: {request.type}")
            
            # 存储任务
            ai_tasks[task_id] = {
                "type": "content_convert",
                "status": "completed",
                "request": request.dict(),
                "result": result
            }
            
            # 返回结果
            return {
                "task_id": task_id,
                "status": "completed",
                "result": result
            }
            
        except Exception as e:
            print(f"内容转换出错: {str(e)}")
            raise HTTPException(status_code=500, detail=f"内容转换失败: {str(e)}")
    
    @staticmethod
    async def recognize_image(file: UploadFile, analyze_type: str = "general") -> Dict[str, Any]:
        """
        调用AI服务进行图像识别
        """
        try:
            # 创建任务ID
            task_id = str(uuid.uuid4())
            
            # 实际项目中，应调用真实的AI API
            # 以下注释部分为实际调用示例代码
            """
            # 读取文件内容
            file_content = await file.read()
            
            # 构建请求
            async with httpx.AsyncClient(timeout=60.0) as client:
                response = await client.post(
                    IMAGE_RECOGNITION_API_URL,
                    headers={"Authorization": f"Bearer {IMAGE_RECOGNITION_API_KEY}"},
                    files={"file": (file.filename, file_content, file.content_type)},
                    data={"analyze_type": analyze_type}
                )
                response.raise_for_status()
                result = response.json()
            """
            
            # 模拟图像识别结果
            if analyze_type == "ocr":
                result = {
                    "text": "这是从图像中提取的文字。实际使用时将返回真实的OCR结果。",
                    "confidence": 0.95
                }
            elif analyze_type == "objects":
                result = {
                    "objects": [
                        {"name": "人", "confidence": 0.98, "box": [10, 10, 100, 200]},
                        {"name": "书", "confidence": 0.85, "box": [150, 50, 200, 100]}
                    ]
                }
            else:  # general
                result = {
                    "description": "这是一张图片的通用描述。实际使用时将返回图像的真实描述。",
                    "tags": ["教育", "课堂", "学习"]
                }
            
            # 存储任务
            ai_tasks[task_id] = {
                "type": "image_recognition",
                "status": "completed",
                "analyze_type": analyze_type,
                "result": result
            }
            
            return {
                "task_id": task_id,
                "status": "completed",
                "result": result
            }
            
        except Exception as e:
            print(f"图像识别出错: {str(e)}")
            raise HTTPException(status_code=500, detail=f"图像识别失败: {str(e)}")
    
    @staticmethod
    async def generate_image(request: ImageGenerationRequest) -> Dict[str, Any]:
        """
        调用AI服务生成图像
        """
        try:
            # 创建任务ID
            task_id = str(uuid.uuid4())
            
            # 实际项目中，应调用真实的AI API
            # 以下注释部分为实际调用示例代码
            """
            async with httpx.AsyncClient(timeout=60.0) as client:
                response = await client.post(
                    IMAGE_GENERATION_API_URL,
                    headers={"Authorization": f"Bearer {IMAGE_GENERATION_API_KEY}"},
                    json={
                        "prompt": request.prompt,
                        "n": request.n,
                        "size": request.size,
                        "style": request.style
                    }
                )
                response.raise_for_status()
                result = response.json()
            """
            
            # 模拟图像生成结果
            images = []
            for i in range(request.n):
                images.append({
                    "url": f"/mock/generated/{task_id}_{i}.png",
                    "width": int(request.size.split("x")[0]),
                    "height": int(request.size.split("x")[1])
                })
                
            result = {
                "created": int(time.time()),
                "images": images
            }
            
            # 存储任务
            ai_tasks[task_id] = {
                "type": "image_generation",
                "status": "completed",
                "request": request.dict(),
                "result": result
            }
            
            return {
                "task_id": task_id,
                "status": "completed",
                "result": result
            }
            
        except Exception as e:
            print(f"图像生成出错: {str(e)}")
            raise HTTPException(status_code=500, detail=f"图像生成失败: {str(e)}")
    
    @staticmethod
    def get_task_status(task_id: str) -> Dict[str, Any]:
        """
        获取任务状态
        """
        if task_id not in ai_tasks:
            raise HTTPException(status_code=404, detail="任务不存在")
        
        return ai_tasks[task_id]
    
    @staticmethod
    def _mock_lesson_plan(
        topic: str,
        grade: str,
        subject: str,
        duration_minutes: int
    ) -> LessonPlan:
        """
        生成模拟教案数据
        """
        # 生成唯一ID
        lesson_id = str(uuid.uuid4())
        
        # 根据学科动态创建教案内容
        subject_content = {
            "语文": {
                "objectives": f"1. 理解{topic}的含义\n2. 掌握相关阅读技巧\n3. 能够进行分析和创作",
                "materials": "教科书、学习单、多媒体设备",
                "assessment": "课堂提问、小组讨论表现、习作评价",
                "sections": [
                    LessonSection(
                        title="导入",
                        content=f"通过提问和图片引入{topic}主题，激发学生兴趣。",
                        duration_minutes=5,
                        activity_type="讲解",
                    ),
                    LessonSection(
                        title="讲解与阅读",
                        content=f"讲解{topic}的主要内容，组织学生进行阅读和思考。",
                        duration_minutes=15,
                        activity_type="阅读理解",
                    ),
                    LessonSection(
                        title="小组讨论",
                        content="学生分组讨论文章的主题和写作特点。",
                        duration_minutes=10,
                        activity_type="小组活动",
                    ),
                    LessonSection(
                        title="练习与应用",
                        content="完成课本习题，进行相关写作练习。",
                        duration_minutes=15,
                        activity_type="个人练习",
                    ),
                    LessonSection(
                        title="总结",
                        content="总结课堂重点，布置课后作业。",
                        duration_minutes=5,
                        activity_type="讲解",
                    ),
                ]
            },
            "数学": {
                "objectives": f"1. 理解{topic}的概念\n2. 掌握相关计算方法\n3. 能够解决实际问题",
                "materials": "教科书、习题集、计算器、几何模型",
                "assessment": "课堂练习、小测验、解题能力",
                "sections": [
                    LessonSection(
                        title="知识回顾",
                        content="复习上节课内容，为新知识做铺垫。",
                        duration_minutes=5,
                        activity_type="讲解",
                    ),
                    LessonSection(
                        title="新概念讲解",
                        content=f"讲解{topic}的核心概念和计算方法。",
                        duration_minutes=15,
                        activity_type="讲解",
                    ),
                    LessonSection(
                        title="例题讲解",
                        content="通过典型例题展示解题思路和方法。",
                        duration_minutes=10,
                        activity_type="讲解与互动",
                    ),
                    LessonSection(
                        title="课堂练习",
                        content="学生独立完成练习题，教师巡视指导。",
                        duration_minutes=15,
                        activity_type="个人练习",
                    ),
                    LessonSection(
                        title="总结与拓展",
                        content="总结本节课知识点，介绍相关应用场景。",
                        duration_minutes=5,
                        activity_type="讲解",
                    ),
                ]
            },
            "英语": {
                "objectives": f"1. 掌握{topic}相关词汇\n2. 提高听说读写能力\n3. 了解相关文化背景",
                "materials": "教科书、音频材料、单词卡片",
                "assessment": "口语表达、听力理解、课文背诵",
                "sections": [
                    LessonSection(
                        title="Warm-up",
                        content="通过简单对话或游戏导入课题。",
                        duration_minutes=5,
                        activity_type="互动",
                    ),
                    LessonSection(
                        title="Vocabulary",
                        content=f"学习{topic}相关的核心词汇。",
                        duration_minutes=10,
                        activity_type="讲解与练习",
                    ),
                    LessonSection(
                        title="Reading & Listening",
                        content="阅读课文并完成听力练习。",
                        duration_minutes=15,
                        activity_type="阅读与听力",
                    ),
                    LessonSection(
                        title="Practice",
                        content="进行口语和写作练习，巩固所学内容。",
                        duration_minutes=15,
                        activity_type="小组活动",
                    ),
                    LessonSection(
                        title="Summary",
                        content="总结本课重点，布置家庭作业。",
                        duration_minutes=5,
                        activity_type="讲解",
                    ),
                ]
            },
        }
        
        # 获取对应学科的内容，如果不存在则使用默认
        subject_data = subject_content.get(subject, {
            "objectives": f"1. 了解{topic}的基本概念\n2. 掌握相关技能\n3. 能够应用到实际中",
            "materials": "教科书、学习材料、辅助工具",
            "assessment": "课堂表现、练习完成情况、课后作业",
            "sections": [
                LessonSection(
                    title="导入",
                    content=f"引入{topic}主题，激发学习兴趣。",
                    duration_minutes=5,
                    activity_type="讲解",
                ),
                LessonSection(
                    title="新知识讲解",
                    content=f"讲解{topic}的核心内容和方法。",
                    duration_minutes=15,
                    activity_type="讲解",
                ),
                LessonSection(
                    title="活动与练习",
                    content="进行相关活动和练习，巩固所学知识。",
                    duration_minutes=15,
                    activity_type="活动与练习",
                ),
                LessonSection(
                    title="总结与反馈",
                    content="总结课堂内容，收集学习反馈。",
                    duration_minutes=5,
                    activity_type="总结",
                ),
            ]
        })
        
        # 创建并返回教案
        return LessonPlan(
            id=lesson_id,
            title=f"{topic} - {grade}{subject}教案",
            subject=subject,
            grade=grade,
            duration_minutes=duration_minutes,
            sections=subject_data["sections"],
            objectives=subject_data["objectives"],
            materials=subject_data["materials"],
            assessment=subject_data["assessment"],
            notes = f"本教案由AI自动生成，基于主题{topic}。",
            created_at=time.strftime("%Y-%m-%d %H:%M:%S"),
        ) 